File: 04_Stack/07_Largest_Rectangle_In_Histogram/0084-largest-rectangle-in-histogram.cpp
/*
Problem: LeetCode 84 - Largest Rectangle in Histogram

Description:
Given an array of integers heights representing the histogram's bar height where the width of each bar is 1, return the area of the largest rectangle in the histogram.

Intuition:
To find the largest rectangle in a histogram, we can utilize a stack to keep track of the indices of increasing heights. By iterating through the histogram, we can calculate the area of each rectangle formed by the heights.

Approach:
1. Initialize a stack to store the indices of increasing heights.
2. Initialize the maximum area to 0.
3. Iterate through each bar in the histogram:
   - While the stack is not empty and the current bar's height is less than the height at the index at the top of the stack:
     - Pop the index from the stack and calculate the area of the rectangle formed by the popped bar.
     - Update the maximum area if the calculated area is greater.
   - Push the current index onto the stack.
4. After iterating through all bars, there might be remaining bars in the stack. Process them similarly to step 3 to calculate the areas and update the maximum area.
5. Return the maximum area.

Time Complexity:
The time complexity is O(n), where n is the number of bars in the histogram. We iterate through each bar once.

Space Complexity:
The space complexity is O(n), where n is the number of bars in the histogram. In the worst case, all bars are stored in the stack.
*/

#include<bits/stdc++.h>
using namespace std;

class Solution {
  public:
    int largestRectangleArea(vector<int> &heights) {
        int n = heights.size();
        stack<int> stack;
        int maxArea = 0;

        for (int i = 0; i <= n; ++i) {
            while (!stack.empty() && (i == n || heights[i] < heights[stack.top()])) {
                int height = heights[stack.top()];
                stack.pop();
                int width = stack.empty() ? i : i - stack.top() - 1;
                maxArea = max(maxArea, height * width);
            }

            stack.push(i);
        }

        return maxArea;
    }
};

File: 04_Stack/07_Largest_Rectangle_In_Histogram/0084-largest-rectangle-in-histogram.rs
/*
Problem: LeetCode 84 - Largest Rectangle in Histogram

Key Idea:
The key idea is to use a stack to keep track of the indices of the histogram bars while iterating through the heights.

Approach:
1. Initialize an empty stack to store indices of histogram bars.
2. Initialize variables `max_area` and `i` to store the maximum area and the current index.
3. Iterate through the heights array:
   - While the stack is not empty and the current height is less than the height at the index at the top of the stack, pop elements from the stack and calculate the area using the height at the popped index and the difference between the current index and the index at the top of the stack. Update `max_area` if the calculated area is greater.
   - Push the current index onto the stack.
4. After the loop, there may be remaining elements in the stack. For each remaining index in the stack, pop it and calculate the area using the height at the popped index and the difference between the current index and the popped index. Update `max_area` if the calculated area is greater.
5. Return `max_area` as the largest rectangle area.

Time Complexity:
O(n), where n is the number of heights in the input array. We perform a single pass through the heights.

Space Complexity:
O(n), as the stack can store at most n indices.
*/

impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let mut stack: Vec<usize> = Vec::new();
        let mut max_area = 0;
        let mut i = 0;

        while i < heights.len() {
            if stack.is_empty() || heights[i] >= heights[*stack.last().unwrap()] {
                stack.push(i);
                i += 1;
            } else {
                let top = stack.pop().unwrap();
                let width = if stack.is_empty() {
                    i
                } else {
                    i - stack.last().unwrap() - 1
                };
                max_area = max_area.max(heights[top] * width as i32);
            }
        }

        while !stack.is_empty() {
            let top = stack.pop().unwrap();
            let width = if stack.is_empty() {
                i
            } else {
                i - stack.last().unwrap() - 1
            };
            max_area = max_area.max(heights[top] * width as i32);
        }

        max_area
    }
}
